knowledge-base/combobox-virtualization-loader.md

```diff
@@ -0,0 +1,159 @@
+                 ItemHeight="30"
+                 PageSize="20"
+                 TextField="@nameof(Person.Name)"
+                 ValueField="@nameof(Person.Id)"
+                 @bind-Value="@SelectedValue"
+                 Filterable="true" FilterOperator="@StringFilterOperator.Contains">
+    <ComboBoxSettings>
+        <ComboBoxPopupSettings Class="example-cb" Height="200px" />
+    </ComboBoxSettings>
+    <NoDataTemplate>
+    </NoDataTemplate>
+    <HeaderTemplate>
+        <TelerikLoader Visible="@IsLoading" />
+    </HeaderTemplate>
+</TelerikComboBox>
+
+@code{
+    bool IsLoading {get;set;} = false;
+    int SelectedValue { get; set; } = 1234; // pre-select an item to showcase the value mapper
+    private TelerikComboBox<Person, int>? ComboBoxRef { get; set; }
+    async Task GetRemoteData(ComboBoxReadEventArgs args)
+    {
+        IsLoading = true;
+        ComboBoxRef?.Refresh();
+        DataEnvelope<Person> result = await MyService.GetItems(args.Request);
+
+        // set the Data and the TotalItems to the current page of data and total number of items
+        args.Data = result.Data;
+        args.Total = result.Total;
+        IsLoading = false;
+        ComboBoxRef?.Refresh();
+    }
+
+    async Task<Person> GetModelFromValue(int selectedValue)
+    {
+        // return a model that matches the selected value so the component can get its text
+        return await MyService.GetItemFromValue(selectedValue);
+    }
+
+    // mimics a real service in terms of API appearance, refactor as necessary for your app
+    public static class MyService
+    {
+        static List<Person> AllData { get; set; }
+
+        public static async Task<DataEnvelope<Person>> GetItems(DataSourceRequest request)
+        {
+            await Task.Delay(3000);
+            if (AllData == null)
+            {
+                AllData = Enumerable.Range(1, 12345).Select(x => new Person { Id = x, Name = $"Name {x}" }).ToList();
+            }
+
+            await Task.Delay(400); // simulate real network and database delays. Remove in a real app
+
+            var result = await AllData.ToDataSourceResultAsync(request);
+            DataEnvelope<Person> dataToReturn = new DataEnvelope<Person>
+            {
+                Data = result.Data.Cast<Person>().ToList(),
+                Total = result.Total
+            };
+
+            return await Task.FromResult(dataToReturn);
+        }
+
+        public static async Task<Person> GetItemFromValue(int selectedValue)
+        {
+            await Task.Delay(400); // simulate real network and database delays. Remove in a real app
+
+            return await Task.FromResult(AllData.FirstOrDefault(x => selectedValue == x.Id));
+        }
+    }
+
+    // used to showcase how you could simplify the return of more than one value from the service
+    public class DataEnvelope<T>
+    {
+        public int Total { get; set; }
+        public List<T> Data { get; set; }
+    }
+
```

Comment:
This part adds an unnecessary and irrelevant complication. Use a simpler example without it.
https://www.telerik.com/blazor-ui/documentation/components/multicolumncombobox/virtualization#remote-data-example